src/io/xpi.js

```diff
@@ -33,7 +33,7 @@ export class Xpi extends IOBase {
     });
   }
 
-  handleEntry(entry, reject) {
+  handleEntry(entry) {
```

Comment:
This method still need to get the `reject` parameter from its callers and to use it to reject the error at line 46 (the DuplicateZipEntry) instead of throwing it as a synchronously raised exception.

The reason is the following:

- `handleEntry` is called from inside an EventEmitter event handler, which is subscribed from inside the Promise returned by the `getFiles` method (in both the "src/io/xpi.js" and "src/io/crx.js" modules)
- the behavior that we expect from the promise returned from `getFiles` is:
  - to be resolved with the list of the file entries once the zipFile and no duplicate entries have been found
  - to be rejected as soon as `handleEntry` is called with a duplicate entry filename

Given that `handleEntry` is called from inside an event handler, if it throws an exception instead of calling `reject`, the raised exception is not going to be translated into a rejection of the Promise returned by `getFiles`, on the contrary it is going to become an "uncaught exception" (which in nodejs can be caught using `process.on("uncaughtException", (err) => { ... });`).

We definitely need to preserve the previous behavior, which is that the rejection will be caught by the Linter class in one of the following `this.handleError(...)` calls:

https://github.com/mozilla/addons-linter/blob/2025658d734ef1153b0bdf3496b1fcc32aff663a/src/linter.js#L479

https://github.com/mozilla/addons-linter/blob/2025658d734ef1153b0bdf3496b1fcc32aff663a/src/linter.js#L499

Also, based on this analysis, the very weird (and concerning) thing is that the different behavior doesn't seem to trigger any test failures, which basically means that something in the current testing strategy is making the tests to behave differently from what is going to happen in the application once executed for real, which is pretty bad and we need to fix it.

After digging a bit into the code and the test suite, I've been able to identify what is affecting the behavior for the tests related to this scenario (related to the DuplicateZipEntry):

In the test cases we are using the `_onEventsSubscribed` parameter of `getFiles` to be able to "fake" the zip file entries when the zipfile listeners have been subscribed, see:

https://github.com/hameleonka/addons-linter/blob/6fd67e426da1130e41feb8c3d5a1c529af71129a/tests/io/test.xpi.js#L215-L219

and 

https://github.com/hameleonka/addons-linter/blob/6fd67e426da1130e41feb8c3d5a1c529af71129a/src/io/xpi.js#L84-L90

But `_onEventsSubscribed` is called synchronously at line 88 of the above snippet, and in the test code linked above the definition of `_onEventsSubscribed` is calling synchronously the `this.handleEntry` method, which had the (unintentional) side-effect that we are experiencing: an Error raised by the new version of the `handleEntry` method is becoming a rejction of the promise returned by `getFiles`... but only in the test cases that are using `_onEventsSubscribed` in this way.

That's very unfortunate and we should absolutely prevent it from happening again (so that we can be more confident that the test suite is going to be able to catch similar regressions), my suggestion is to change the way we call the `_onEventSubscribed` callback, so that it is not called synchronously from inside the Promise, and better match what happens when `this.handleEntry` is called from inside the zipfile listeners, e.g. by chaging it into:

```js
      if (_onEventsSubscribed) {
        // Run optional callback when we know the event handlers
        // have been inited. Useful for testing.
        if (typeof _onEventsSubscribed === 'function') {
          Promise.resolve().then(() => _onEventsSubscribed(reject));
        }
      }
```

Reply:
@rpl as you suggested in order to see the difference between throw error and reject I put throw error in the beginning of handleEntry function and saw that with reject this error is handled with handleError function and without reject node process just fails. Thanks for your help!